source/common/upstream/cluster_manager_impl.cc

```diff
@@ -281,23 +308,31 @@ void ClusterManagerImpl::ThreadLocalClusterManagerImpl::shutdown() {
 ClusterManagerImpl::ThreadLocalClusterManagerImpl::ClusterEntry::ClusterEntry(
     ThreadLocalClusterManagerImpl& parent, const Cluster& cluster, Runtime::Loader& runtime,
     Runtime::RandomGenerator& random, Stats::Store& stats_store, Event::Dispatcher& dispatcher,
-    const std::string& local_zone_name, const std::string& local_address)
+    const std::string& local_zone_name, const std::string& local_address,
+    const ClusterEntry* local_cluster)
     : parent_(parent), primary_cluster_(cluster),
       http_async_client_(
           cluster, stats_store, dispatcher, local_zone_name, parent.parent_, runtime, random,
           Router::ShadowWriterPtr{new Router::ShadowWriterImpl(parent.parent_)}, local_address) {
 
+  const HostSet* local_host_set = nullptr;
```

Comment:
unless you are going to store const ClusterEntry* somewhere just pass const HostSet* to this function then you avoid doing the same if check again here as you do above.